lib/internal/modules/esm/import_map.js

```diff
@@ -0,0 +1,107 @@
+'use strict';
+const { isURL, URL } = require('internal/url');
+const { ObjectEntries, ObjectKeys, SafeMap, ArrayIsArray } = primordials;
+const { codes: { ERR_INVALID_IMPORT_MAP } } = require('internal/errors');
+
+class ImportMap {
+  #baseURL;
+  imports = new SafeMap();
+  scopes = new SafeMap();
+
+  constructor(raw, baseURL) {
+    this.#baseURL = baseURL;
+    processImportMap(this, this.#baseURL, raw);
+  }
+
+  get baseURL() {
+    return this.#baseURL;
+  }
+
+  resolve(specifier, parentURL = this.baseURL) {
+    // Process scopes
+    for (const { 0: prefix, 1: mapping } of this.scopes) {
+      let mappedSpecifier = mapping.get(specifier);
+      if (parentURL.pathname.startsWith(prefix.pathname) && mappedSpecifier) {
+        if (!isURL(mappedSpecifier)) {
+          mappedSpecifier = new URL(mappedSpecifier, this.baseURL);
+          mapping.set(specifier, mappedSpecifier);
+        }
+        specifier = mappedSpecifier;
+        break;
+      }
+    }
+
+    let spec = specifier;
+    if (isURL(specifier)) {
+      spec = specifier.pathname;
+    }
+    let importMapping = this.imports.get(spec);
+    if (importMapping) {
+      if (!isURL(importMapping)) {
+        importMapping = new URL(importMapping, this.baseURL);
+        this.imports.set(spec, importMapping);
+      }
+      return importMapping;
+    }
+
+    return specifier;
+  }
+}
+
+function processImportMap(importMap, baseURL, raw) {
+  // Validation and normalization
+  if (typeof raw.imports !== 'object' || ArrayIsArray(raw.imports)) {
```

Comment:
We need to account that `typeof null === 'object'`
```suggestion
  if (raw.imports == null || typeof raw.imports !== 'object' || ArrayIsArray(raw.imports)) {
```

Reply:
Yeah I had originally just done a basic check, but as I was checking this against the spec last night I noticed it does say to throw a type error if it is not an ordered map. I didn't update because I wanted to get the PR open, but I think I need to make two changes here:

1. check for `null`
2. create another error for invalid import map which extends `TypeError` for this case.

Reply:
We have generic helpers like `validateString`, if there isn’t a `validatePlainObject` it might be worth creating one. What do you think @aduh95?

Reply:
We can address the need for helpers here later I think. Honestly I would even be fine doing that as a follow up PR. Calling this resolved as I made the change locally and will be pushing it today.